### ISSUE TITLE
Prod - Enable S3 sync to deploy only changed files

### ISSUE THREAD

**Issue by laineymajor** [department-of-veterans-affairs/va.gov-team#35198]
Process for this is complete (currently in staging). Change in CI to include PROD.

**Comment by theodur** [comment]
One issue that we may run into is a race condition between the S3 production sync for changed app deploys, and the daily production deploy. We need to account for when we have both a changed app prod deploy and full production deploy occurring at the same time. If a changed app build commit gets merged into `main` during the 60-minute wait time during the deploy, the latter changed app build commit will be deployed before the full prod deploy completes, meaning the prior commit that was used for the daily deploy will be deployed, overriding the latter commit's deployment.

## Possible solutions

### Same deployment workflow for changed app prod deploys and full prod deploys
One way to avoid potential race conditions is to use the same deployment job for both types of production deploys. This approach ensures that the production S3 buckets are never written to at the same time. There would need to be a script in CI that determines when to add a new changed app production deploy to the workflow queue. We would only want to add a new changed app production deploy to the queue when a completed CI workflow run in `main` is from a commit that's ahead of the latest deployed commit, or scheduled deployed commit, i.e. we wouldn't want triggered re-runs from previous changed app commits to add deployments to the deployment queue, because this would unintentionally revert the deployed code for that changed app build

**Comment by theodur** [comment]
# Problem Statement
There are currently 2 main issues that are blocking the continuous production deployment of applications on the [allowlist](https://github.com/department-of-veterans-affairs/vets-website/blob/main/config/changed-apps-build.json).

1. A race condition that will exist between the daily production deploy and changed app production deploys
2. Changed app production deploys may be built and tested with recent updates to global assets that aren't in the S3 production bucket at the time of their deployments

To get a better understanding of the two blocking issues, it's helpful to look at the current infrastructure of changed app builds. Currently, when a commit is pushed to a remote feature branch in the `vets-website` repo, the CI workflow in GHA is triggered. The CI workflow determines whether the code changes in the branch are solely from applications on the allowlist. If so, the workflow will run a build that only generates the assets of the allowed application(s) that were modified in the branch, along with any global assets needed for the application(s) to run. Here's a [Tarbell](https://s3-us-gov-west-1.amazonaws.com/vetsgov-website-builds-s3-upload/c1ac734eaa93d23e1108790f5b53ae85bb9b68a3/vagovstaging.tar.bz2) of assets that were generated in a changed app build (assets are in the `generated` directory). Subsequently, unit tests and Cypress tests for the modified allowed applications are also run in the CI workflow.

The problem arises when feature branches are merged into the `main` branch. CI for the `main` branch works similarly as it does for feature branches, in that the workflow only runs a build for the allowed apps that were modified in the merged commit. Unit tests and Cypress tests for the modified allowed apps are also run in the workflow. If the merged commit contains code changes outside the applications on the allowlist, a full build is run in the CI workflow, along with all unit and Cypress tests. One main distinction is that the `main` branch deploys commits to the dev and staging environments at the end of the CI workflow. This deployment happens for both commits with full builds, and commits with changed app builds. For commits with full builds, all assets generated from the build are synced to the `generated` directory in the appropriate S3 bucket for the environment. The assets of each application are stored together in the `generated` directory, i.e. `vaos.entry.js` and `facilities.entry.js` are stored alongside each other in the directory. For commits with changed app builds, only the assets of the modified applications are synced to the S3 bucket for the appropriate environment, meaning if changes to the VAOS app are merged into the `main` branch for example, only `vaos.entry.js`, `vaos.css`, etc. are synced to the `generated` directory in the S3 bucket, leaving other assets untouched.

For commits in `main` with changed app builds, there's an assumption being made that the global assets generated in the changed app build are the same as the global assets that are in the S3 bucket. This may not always be the case, which could possible cause a changed app deployment to break the application. For example, if a full build commit making substantial changes that affect global assets gets merged into `main`, and a changed app build commit is merged afterwards, it's not guaranteed that the global assets in the S3 bucket have been updated with the changes in the previous full build commit. The changed app build commit would still be built using those substantial global asset changes in CI, but because those global asset updates may not be in the S3 bucket, the changed app's synced assets might not work correctly when synced to the S3 buckets.

# Proposed Solution

## Overview
To avoid changed app assets being dependent on global asset changes that may not be synced to the S3 bucket, one approach is to change the structure of the `generated` directory, so that the assets built and tested in changed app build commits are the assets being for the applications on the site. Another reason why this route is beneficial is because Cypress tests in changed app build commits use the assets generated from the changed app build for testing, rather than testing with the changed app's assets synced with the global assets from the S3 bucket. This means there's no way of knowing if the changed app's assets synced with the live global assets would pass Cypress tests or not. Using the actual assets that have been tested in production is a safer approach that ensures the functionality of allowed applications.

This solution would entail having subdirectories in the `generated` directory that are for changed app build assets, i.e. assets deployed from VAOS changed app builds would be stored in `generated/vaos`, etc., while the assets of applications that aren't on the allowlist will be kept in the root of the `generated` directory as they are now. The benefit of this is that it separates full builds from changed app builds, which means the builds and deployments can be truly isolated. This removes any race conditions we would have, because the isolated apps wouldn't be part of the daily deploy. This is also probably the only way we can remove allowed apps from the daily production deploy.

## Updating the content-build
Taking this route would require updating the content build to know when to use a subdirectory in the `generated` directory for an app's page. One way this could be done is by adding an attribute to the entry of allowed apps in the [registry](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) that would delineate allowed apps and non-allowed apps. An attribute like `isAllowedApp` being set to `true` could be used to tell the content-build that it should use a subdirectory for referencing assets on that app's page. There are some caveats that need to be addressed, such as how to change the bucket path for changed apps without breaking the app for a certain time.

## Updating CI for different build types
Having separate subdirectories for allowed apps means they can't be built with the full build. This presents some challenges in CI for commits that make global (non-app) changes in the repo. CI would need to run the full build with apps that aren't on the allowlist, along with separate changed app builds for apps that are on the allowlist. This also means that each of those builds need to have their respective Cypress and unit tests ran separately. Yarn workspaces may be able to help with this. There is the [yarn workspaces foreach](https://yarnpkg.com/cli/workspaces/foreach) command that we could use to run the build for each allowed app. This would mean that each allowed app's `package.json` would have to have a `build` script that would run the build for its app. The `foreach` command also has a `--parallel` option that can run the specified command in each workspace in parallel. The same could be done for Cypress tests and unit tests.

## Determining production deployment process for allowed apps
Since apps on the allowlist wouldn't be part of the daily production deploy in this solution, we would need to determine what automated deployments would look like for allowed apps. There are many options in determining this, since allowed apps wouldn't have to be concerned with colliding with the daily production deploy. We could have a dedicated GHA workflow for all allowed app deployments that runs on a specific cron. This workflow would deploy allowed apps separately. Another option if we wanted to provide complete autonomy would be to use [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) in GHA. This could allow each allowed app to have their own production deployment workflow, and set their own cron, while reusing a standard workflow that we create and maintain. There are many options for how we can define automated prod deployments for allowed apps, these are just a few possible routes.

**Comment by timwright12** [comment]
Overall, it feels like we're at a bit of a crossroads with this initiative. My comments here are:

- If we're changing the directory paths, can move allowed apps from `/generated` to `/dist` to follow a more industry standard?
- The content-build solution outlined really feels like we're piling tech debt on top of tech debt to me (adding conditionals and more data into the registry that already exists on the allowlist in vets-website). I'd like to figure out a cleaner way around that like what we spoke about in slack by moving that path ownship over to `vets-website`

> We could have a dedicated GHA workflow for all allowed app deployments that runs on a specific cron. 

Why would we use a cron here if they trigger for a deploy isn't time-base? The overall point here, I believe, is to let teams deploy on merge to `main`

**Comment by theodur** [comment]
@timwright12 
> If we're changing the directory paths, can move allowed apps from /generated to /dist to follow a more industry standard?

Sure, we could change it to `dist` if that seems more appropriate. I only mentioned `generated` since that's the directory we've been using. The important part about this solution is to have application's have their own subdirectories for their assets, wherever that is.

> The content-build solution outlined really feels like we're piling tech debt on top of tech debt to me (adding conditionals and more data into the registry that already exists on the allowlist in vets-website). I'd like to figure out a cleaner way around that like what we spoke about in slack by moving that path ownership over to vets-website

Agreed, Ideally I wouldn't like to make substantial changes to the content build, especially since it's in the process of being migrated from. That was just one option that came to mind, but there's room for discovery for other solutions for that issue, like the one you mentioned

> Why would we use a cron here if the trigger for a deploy isn't time-base? The overall point here, I believe, is to let teams deploy on merge to main

The thought process was that we may still want some type of automated deploy for allowed apps. If there are security updates that need to go out in the repo, there's not a guarantee that app teams will merge a new commit to `main` anytime soon, which means they could have vulnerable code in production until they merge a new commit to `main`, since they wouldn't be part of the daily deployment. Also, this could solve the issue we've been discussing of teams not wanting to deploy to production as frequent. This part of the solution isn't absolutely necessary since the main goal is to do daily production deploys on `main` merges, just mentioning this since a we'd have the option to do that with this approach. But we'd still need a workflow so they could deploy manually if needed

**Comment by theodur** [comment]
After creating an RFC for the problem mentioned above, we ultimately decided that the syncing of application bundles and global bundles in the deployment bucket would rarely cause issues, since global bundles don't often change. A more descriptive [explanation](https://github.com/department-of-veterans-affairs/va.gov-platform-architecture/pull/32#issuecomment-1183634886) was added to the RFC. As a result, the only issue we have to solve is a race condition between changed app deploys and the daily production deploy. We won't need to update the build directory structure for this, because we can add logic to the `Deploy` job in the CI workflow to pause changed app deploys when needed.

Enabling the S3 sync for prod builds in changed app commits should be straight forward after adding the necessary logic to the `Deploy` job to handle when to pause changed app deploys. This will most likely be a node script that makes a call to the GitHub API to determine if there's a daily production deploy workflow run in progress.

**Comment by theodur** [comment]
Draft PR for enabling S3 syncing for prod: https://github.com/department-of-veterans-affairs/vets-website/pull/21716

**Comment by theodur** [comment]
I marked the [PR](https://github.com/department-of-veterans-affairs/vets-website/pull/21716) as ready for review. Will merge once approved. This PR won't turn on continuous production deployments, it only adds the logic for it. Another PR will be opened for setting the `DEPLOY_TO_PRODUCTION` variable to true, which will turn on continuous prod deployments

**Comment by theodur** [comment]
The [PR](https://github.com/department-of-veterans-affairs/vets-website/pull/21716) for adding logic to handle the race condition has been merged. I will monitor CI to make sure everything is working correctly

**Comment by theodur** [comment]
The [PR](https://github.com/department-of-veterans-affairs/vets-website/pull/21996) to add an option for toggling continuous deployment has been updated based on the decision for apps on the allowlist to be opeted in to continuous deployment by default

**Comment by theodur** [comment]
The PR for adding an option to disable CD has been merged. Now all that's left is waiting for teams to decide whether they want to disable CD for their apps on the allowlist. Once that is complete and the necessary apps have CD disabled, we can open a PR for turning on continuous deployment in the CI workflow.

**Comment by theodur** [comment]
The [PR](https://app.zenhub.com/workspace/o/department-of-veterans-affairs/vets-website/issues/22168) for turning on continuous deployment has been merged ✅ I will begin going through items in the [test plan](https://docs.google.com/spreadsheets/d/1skAc2yGVS_iuXv8K5sh85HId93Sw-39wuVC7s3d2TuM/edit?usp=sharing) as continuous deployments come

**Comment by theodur** [comment]
After monitoring continuous deployment, everything seems to be working as expected
